server/app/controllers/applicant/ApplicantProgramReviewController.java

```diff
@@ -106,12 +117,41 @@ public CompletionStage<Result> submit(Request request, long applicantId, long pr
                 if (cause instanceof SecurityException) {
                   return unauthorized();
                 }
+                if (cause instanceof ApplicationOutOfDateException) {
+                  Call reviewPage =
+                      routes.ApplicantProgramReviewController.review(applicantId, programId);
+                  return redirect(reviewPage).flashing("error", cause.getMessage());
+                }
                 throw new RuntimeException(cause);
               }
               throw new RuntimeException(ex);
             });
   }
 
+  /**
+   * Validates that the application is complete and correct to submit.
+   *
+   * <p>An application may be submitted but incomplete for a variety of reason:
+   *
+   * <ul>
+   *   <li>The application view with submit button contains stale data that has changed visibility
+   *       or eligibility conditions that result in the application being incomplete or ineligible.
+   * </ul>
+   *
+   * @return a {@link ApplicationOutOfDateException} wrapped in a failed future with a user visible
+   *     message for the issue.
+   */
+  private CompletableFuture<Void> validateApplication(
+      ReadOnlyApplicantProgramService roApplicantProgramService, Messages messages) {
+    // Check that all blocks have been answered.
+    if (!roApplicantProgramService.getFirstIncompleteBlockExcludingStatic().isEmpty()) {
+      return CompletableFuture.failedFuture(
+          new ApplicationOutOfDateException(
+              messages.at(MessageKey.TOAST_APPLICATION_OUT_OF_DATE.getKeyName())));
+    }
+    return CompletableFuture.completedFuture(null);
```

Comment:
suggestion (blocking): this logic should be encapsulated in a service rather than a controller (remember the litmus test: the same logic should be reusable if we added another endpoint to do this via API). Suggest moving this logic into either the implementation of `applicantService.submitApplication` or as a new method on `AppplicantService`

Reply:
done